### PR TITLE
[inductor] Add support for tl.make_block_ptr

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116360
* __->__ #116079

On A100 this is a small regression:
image

So I will leave it disabled by default.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -491,6 +491,7 @@ class DeferredLine(DeferredLineBase):
     def __init__(self, name, line):
         super().__init__(line)
         self.name = name
+        assert not isinstance(line, DeferredLineBase)
```
Comment (shunting314) [inline torch/_inductor/codegen/common.py:494]: nit: should this be done in the base class?

**Review by shunting314 — APPROVED** [review]

**Comment by shunting314** [comment]
Based on the perf results, block ptr mainly benefits H100?

**Comment by jansel** [comment]
> Based on the perf results, block ptr mainly benefits H100?

There was a discussion on slack saying block_ptrs won't help with TMAs on H100 either, and are the wrong abstraction. 
https://triton-lang.slack.com/archives/C01LY4FJL56/p1702950152240829?thread_ts=1702944987.981749&cid=C01LY4FJL56

Though it could help with other hardware backends.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but the new test is failing on ROCm' -c nosignal

Here is an example failure https://hud.pytorch.org/pytorch/pytorch/commit/b8374314cc717a3f6825d876072f851a9ebef228.  This is a landrace of sort because `inductor/test_torchinductor` starts to run on ROCm only after https://github.com/pytorch/pytorch/pull/115425

**Comment by jithunnair-amd** [comment]
@jataylo @pragupta Heads up please, this PR was reverted because it broke ROCm. Can you please assess it and see if it would need a triton update, or is it something we can submit a patch for? Last resort would be to skip any associated unit tests if we need more time to figure out a proper resolution.

**Comment by jansel** [comment]
Updated PR to add a `@skipIfRocm`.

**Pull request closed without merging**